packages/aws-cdk/lib/cdk-toolkit.ts

```diff
@@ -447,6 +517,70 @@ export class CdkToolkit {
   }
+
+  /**
+   * Returns all NestedStack templates
+   */
+  private getNestedStacks(stack: cxapi.CloudFormationStackArtifact): Array<{parentPath?: string, parentResourceId: string, template: any}> {
+
+    const ret = new Array<{ parentPath?: string, parentResourceId: string, template: any }>();
+    const resources = [{resources: stack.template?.Resources, parentStackResources: {} as any, parentStackId: stack.stackName}];
+    const nestedStackAssets = stack.assets.filter((ast) => {
+      return ast.path && ast.path.includes('nested.template.json');
+    });
+
+    while (nestedStackAssets.length > 0) {
+      for (const running of resources) {
+        for (const key in running.resources) {
+          if (running.resources[key].Type === 'AWS::CloudFormation::Stack') {
+            const s3Key = this.findS3VersionKeyForNestedStackResource(running, key, resources);
+            for (const ast of nestedStackAssets) {
+              if (!!s3Key && (ast as any).s3KeyParameter === s3Key) {
+                const template = deserializeStructure(fs.readFileSync(path.join(stack.assembly.directory, ast.path)).toString('utf-8'));
+                ret.push({parentPath: template.Metadata?.[cxapi.PATH_METADATA_KEY], parentResourceId: key, template});
+                resources.push({
+                  resources: template.Resources,
+                  parentStackResources: running.resources,
+                  parentStackId: key,
+                });
+                nestedStackAssets.splice(nestedStackAssets.indexOf(ast), 1);
+                break;
+              }
+            }
+          }
+        }
+      }
+    }
+    return ret;
+  }
+
+  private findS3VersionKeyForNestedStackResource(
+    running: { resources: any; parentStackId: string; parentStackResources: any },
+    key: string,
+    resources: Array<{ resources: any; parentStackId: string; parentStackResources: any }>) {
+    const url = JSON.stringify(running.resources[key].Properties?.TemplateURL);
+    let s3Key = url.match(/\w+S3VersionKey\w+/g)?.[0];
+    if (!!s3Key && s3Key.includes('referenceto')) {
+      if (Object.keys(running.parentStackResources[running.parentStackId].Properties?.Parameters || {}).includes(s3Key)) {
+        s3Key = running.parentStackResources[running.parentStackId].Properties?.Parameters[s3Key].Ref;
+      }
+      if (!!s3Key && s3Key.includes('referenceto')) {
+        let resUtil = [...resources];
+        while (resUtil.length > 0) {
+          const res = resUtil.pop() as any;
+          if (Object.keys(res.parentStackResources[res.parentStackId]?.Properties?.Parameters || {}).includes(s3Key!)) {
+            s3Key = res.parentStackResources[res.parentStackId].Properties?.Parameters[s3Key!].Ref;
+            if (s3Key!.includes('referenceto')) {
+              resUtil = [...resources];
+            } else {
+              break;
+            }
+          }
+        }
+      }
+    }
+    return s3Key;
```

Comment:
I think it's worth breaking the nested stack parsing out into it's own class so it's more unit-testable. Relying on integ tests has put us in a place that we would like to get out of by raising the unit test coverage.